### PR TITLE
Flush instruction cache after thunk pool allocation

### LINKED ISSUE #74710: System.Net.* and System.Security.* crash on NativeAOT arm64
Frequency as of:
- 9/9 - [last 30 days of System.Net.Security.Tests](https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E30+definition%3Aruntime-extra-platforms+name%3A%22system.net.security.tests+work+item%22) (with Core Dump) - all arm64-NativeAOT only
- 9/9 - [last 30 days of System.Net.WebSockets.Client.Tests](https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E30+definition%3Aruntime-extra-platforms+name%3A%22system.net.websockets.client.tests+work+item%22) (with Core Dump) - all arm64-NativeAOT only
- 9/9 - [last 30 days of System.Net.Http.Functional.Tests](https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E30+definition%3Aruntime-extra-platforms+name%3A%22System.Net.Http.Functional.Tests+work+item%22) (with Core Dump) - mostly arm64-NativeAOT with 2 exceptions crossed out in the table
- 9/9 - [last 30 days of System.Security.Cryptography.Tests](https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E30+definition%3Aruntime-extra-platforms+name%3A%22system.security.cryptography.tests+work+item%22) (with Core Dump) - all arm64-NativeAOT only

Day | Test suite | Run | OS | Details
-- | -- | -- | -- | --
9/8 | System.Security.Cryptography.Tests | PR [9648](https://runfo.azurewebsites.net/view/build/?number=9648) - PR #75306 (main) | net7.0-Linux-Release-arm64-NativeAOT_Release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-75306-merge-171bd00681b64e6184/System.Security.Cryptography.Tests/1/core.1001.23?%3Fhelixlogtype%3Dresult) - SIGILL
9/8 | System.Security.Cryptography.Tests | PR [9492](https://runfo.azurewebsites.net/view/build/?number=9492) - PR #75213 (main) | net7.0-Linux-Release-arm64-NativeAOT_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-75213-merge-42ee583a13b04dd886/System.Security.Cryptography.Tests/1/core.1001.25?%3Fhelixlogtype%3Dresult) - SIGILL
9/4 | System.Net.WebSockets.Client.Tests | Rolling run [4489](https://runfo.azurewebsites.net/view/build/?number=4489) (main) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-f030648dc31e43a9b6/System.Net.WebSockets.Client.Tests/1/core.1001.22?%3Fhelixlogtype%3Dresult) - SIGILL
9/2 | System.Net.Http.Functional.Tests | Rolling run [3461](https://runfo.azurewebsites.net/view/build/?number=3461) (main) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-329b3e5ed90e40608d/System.Net.Http.Functional.Tests/1/core.1001.25?%3Fhelixlogtype%3Dresult) - SIGILL
9/1 | System.Net.Http.Functional.Tests | Rolling run [1981002](https://runfo.azurewebsites.net/view/build/?number=1981002) (7.0-rc1) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-rc1-48b6e9caf6244ae3b9/System.Net.Http.Functional.Tests/1/core.1001.25?%3Fhelixlogtype%3Dresult) - SIGILL
8/28 | System.Net.Security.Tests | Rolling run [1972481](https://runfo.azurewebsites.net/view/build/?number=1972481) (main) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-3a60d3185e10463987/System.Net.Security.Tests/1/core.1001.22?%3Fhelixlogtype%3Dresult) - SIGILL
~8/28~ | ~System.Net.Http.Functional.Tests~ | ~Rolling run [1972750](https://runfo.azurewebsites.net/view/build/?number=1972750)~ | ~net7.0-windows-Release-**x64-NativeAOT**_Release-Windows.81.Amd64.Open~ | ~[Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-rc1-7fcad81b3d5846afb8/System.Net.Http.Functional.Tests/1/System.Net.Http.Functional.Tests.exe.4024.dmp?%3Fhelixlogtype%3Dresult) - Unhandled Exception: System.NullReferenceException~
8/28 | System.Net.Http.Functional.Tests | Rolling run [1972413](https://runfo.azurewebsites.net/view/build/?number=1972413) (7.0) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-238c67e5cc7c4be6bb/System.Net.Http.Functional.Tests/1/core.1001.26?%3Fhelixlogtype%3Dresult) - SIGILL
8/27 | System.Net.WebSockets.Client.Tests | Rolling run [1972127](https://runfo.azurewebsites.net/view/build/?number=1972127) (main) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-56d30ef28a0d4fc187/System.Net.WebSockets.Client.Tests/1/core.1001.23?%3Fhelixlogtype%3Dresult) - SIGILL
8/26 | System.Net.Security.Tests | Rolling run [1969936](https://runfo.azurewebsites.net/view/build/?number=1969936) (7.0) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-ec0251bfa63f4e6385/System.Net.Security.Tests/1/core.1001.26?%3Fhelixlogtype%3Dresult) - SIGILL
~8/25~ | ~System.Net.Http.Functional.Tests~ | ~Merged PR [1966082](https://runfo.azurewebsites.net/view/build/?number=1966082)~ | ~net7.0-windows-Release-**x64-NativeAOT**_Release-Windows.81.Amd64.Open~ | ~[Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-74556-merge-c22b6586a4bf4bdfbc/System.Net.Http.Functional.Tests/1/System.Net.Http.Functional.Tests.exe.4156.dmp?%3Fhelixlogtype%3Dresult) - [FAIL] System.Net.Http.Functional.Tests.SyncHttpHandlerTest_AutoRedirect.AllowAutoRedirect_True_ValidateNewMethodUsedOnRedirection~
8/21 | System.Net.Http.Functional.Tests | Rolling run [1957446](https://runfo.azurewebsites.net/view/build/?number=1957446) (7.0-rc1) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-rc1-b49b757c823a457a9b/System.Net.Http.Functional.Tests/1/core.1001.25?%3Fhelixlogtype%3Dresult) - SIGILL
8/20 | System.Net.Security.Tests | Rolling run [1956604](https://runfo.azurewebsites.net/view/build/?number=1956604) (main) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-b0836dc6bf32443dac/System.Net.Security.Tests/1/core.1001.25?%3Fhelixlogtype%3Dresult) - SIGILL
8/20 | System.Net.WebSockets.Client.Tests | Rolling run [1955765](https://runfo.azurewebsites.net/view/build/?number=1955765) (7.0-rc1) | net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-rc1-a399a3bfe7f04a658c/System.Net.WebSockets.Client.Tests/1/core.1001.26?%3Fhelixlogtype%3Dresult) - SIGILL

Not quite blocking CI yet, but close ...

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #74710

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp [hunk 1]
```diff
@@ -832,6 +832,34 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalVirtualProtect(_In_ void* pAddre
     return mprotect(pPageStart, memSize, unixProtect) == 0;
 }
 
+REDHAWK_PALEXPORT void PalFlushInstructionCache(_In_ void* pAddress, size_t size)
+{
+#ifndef HOST_ARM
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:837]: copy & paste from coreclr pal

**Comment by filipnavara** [comment]
I will rebase #75264 if this lands first.

**Review by filipnavara — APPROVED** [review]

**Comment by filipnavara** [comment]
I noticed the same issue when fixing the code for macOS ARM64 so I am actually glad to have a more complete fix. Not being too familiar with the ARM64 architecture I have a follow-up question. Since the data pages are written separately (from managed code) is there any need to flush data cache for these pages too? I originally came to the conclusion that it should not be necessary but since you are way more familiar with the code I assume you'll know better.

**Comment by jkotas** [comment]
The data pages will get written to when the marshalled delegate is created. Locks taken during that process should take care of the data catches flushing.

File: src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp (hunk 1 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:837]: Is this needed on macOS?

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:837]: HOST_ARM is 32bit ?

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:837]: Looks like HOST_ARM and HOST_ARM64 are mutually exclusive.

https://github.com/dotnet/runtime/blob/76533d4f528a8c51457d9dd0c438f3babbd7733b/src/coreclr/jit/jit.h#L69

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp]: Yes, host arm is 32-bit.

I have updated the ifdef for clarity.

**Comment by VSadov** [comment]
We probably see Windows issues due to this as well, perhaps not as often for some reason.

**Comment by jkotas** [comment]
> We probably see Windows issues due to this as well, perhaps not as often for some reason.

Yes. 

This path is only hit by marshalled delegates. We tried to rid of marshalled delegates from everywhere in product code and replace them with function pointers. We have hit the bug in libraries test only because this one marshalled delegate was missed during the conversion. I am going to submit a separate PR to convert this marshalled delegate to a function pointer.

**Review by VSadov — APPROVED** [review]
LGTM! It was probably hard to figure the root cause.

**Comment by VSadov** [comment]
With this and the dealings with JIT memory in the other PR, I wonder - is there is a configuration where NativeAOT does not use executable memory?  (even if with some tradeoffs)

**Comment by jkotas** [comment]
> is there a configuration where NativeAOT does not use executable memory? (even if with some tradeoffs)

We have a configuration under ifdef (that is currently disabled) that compiles these entrypoints into the binary as static code, and then creates copies of this static code if necessary using file mapping. It is not very portable and makes the binary bigger. We can make it work if somebody asks for it.

**Comment by MichalPetryka** [comment]
> > is there a configuration where NativeAOT does not use executable memory? (even if with some tradeoffs)
> 
> We have a configuration under ifdef (that is currently disabled) that compiles these entrypoints into the binary as static code, and then creates copies of this static code if necessary using file mapping. It is not very portable and makes the binary bigger. We can make it work if somebody asks for it.

I'd assume that mode is used by the teams that made NativeAOT work on consoles that are AOT only.

**Comment by jkotas** [comment]
> I'd assume that mode is used by the teams that made NativeAOT work on consoles that are AOT only.

Yep. Or they just avoid marshalled delegates to avoid the problem all-together.

**Comment by bartonjs** [comment]
> This path is only hit by marshalled delegates. We tried to rid of marshalled delegates from everywhere in product code and replace them with function pointers. 

Does that mean there's some test coverage missing somewhere for the scenario?  If it took us two weeks to figure out what was wrong I'd hate to think what it would mean for the average customer.

**Comment by jkotas** [comment]
> Does that mean there's some test coverage missing somewhere for the scenario? 

We do have a test coverage. The crash is caused by a race condition with processor instruction cache. It is impossible to write tests that would reliably catch these types of race conditions.

**Pull request merged**